### PR TITLE
Factor out a method to parse termination messages

### PR DESCRIPTION
**Pull request by ImJasonH**
This reduces duplication, and allows us to make future changes to how
termination messages are written in lock-step with how they're parsed.

cc @othomann @chengjingtao

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [y] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [n/a] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [y] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

### CODE REVIEW

**Comment by othomann** [comment]
@ImJasonH I think the ParseMessage function needs to handle an empty string, this might explain the problem in the integration test. So the parse tests should have a test with an empty string.

**Comment by ImJasonH** [comment]
> @ImJasonH I think the ParseMessage function needs to handle an empty string

Thanks, that was definitely the bug. Handled, and tested

File: go.mod [hunk 1]
```diff
@@ -4,6 +4,7 @@ go 1.13
 
 require (
 	cloud.google.com/go v0.47.0 // indirect
+	cloud.google.com/go/storage v1.0.0
```
Comment (bobcatfish) [inline go.mod]: :eyes:

File: pkg/reconciler/taskrun/taskrun.go [hunk 2]
```diff
@@ -397,28 +399,21 @@ func (c *Reconciler) handlePodCreationError(tr *v1alpha1.TaskRun, err error) {
...
-				if msg != "" {
-					if err := updateTaskRunStatusWithResourceResult(taskRun, []byte(msg)); err != nil {
-						logger.Infof("No resource result from %s for %s/%s: %s", cs.Name, taskRun.Name, taskRun.Namespace, err)
-					}
+				r, err := termination.ParseMessage(msg)
+				if err != nil {
+					return err
 				}
+				// TODO(jasonhall): If two results are defined
+				// for the same resource, take the last one.
```
Comment (bobcatfish) [inline pkg/reconciler/taskrun/taskrun.go]: for TODOs can we reference an issue number or not include them at all? that way we can have a way to check if it's been followed up on later and also it doesnt have to be you necessarily that fixes it

File: go.mod (hunk 1 above)
Comment (ImJasonH) [reply go.mod]: 🤷‍♂ 

I didn't make this change on purpose, Go must have added it.

File: pkg/reconciler/taskrun/taskrun.go [hunk 3]
```diff
@@ -397,28 +399,21 @@ func (c *Reconciler) handlePodCreationError(tr *v1alpha1.TaskRun, err error) {
...
+		for _, cs := range podStatus.ContainerStatuses {
 			if cs.State.Terminated != nil {
 				msg := cs.State.Terminated.Message
-				if msg != "" {
-					if err := updateTaskRunStatusWithResourceResult(taskRun, []byte(msg)); err != nil {
-						logger.Infof("No resource result from %s for %s/%s: %s", cs.Name, taskRun.Name, taskRun.Namespace, err)
-					}
+				r, err := termination.ParseMessage(msg)
+				if err != nil {
+					return err
```
Comment (bobcatfish) [inline pkg/reconciler/taskrun/taskrun.go]: i'm curious why in this case and in the `return err` above you didn't want to add some more context to the error with fmt.errorf or similiar? i find i helps in debugging to add a bit of context along the way

File: pkg/reconciler/taskrun/taskrun_test.go [hunk 4]
```diff
@@ -1515,42 +1515,22 @@ func TestReconcileCloudEvents(t *testing.T) {
...
-func TestUpdateTaskRunStatus_withValidJson(t *testing.T) {
+func TestUpdateTaskRunResourceResult(t *testing.T) {
 	for _, c := range []struct {
-		desc    string
-		podLog  []byte
-		taskRun *v1alpha1.TaskRun
-		want    []v1alpha1.PipelineResourceResult
+		desc          string
+		podStatus     corev1.PodStatus
+		taskRunStatus *v1alpha1.TaskRunStatus
```
Comment (bobcatfish) [inline pkg/reconciler/taskrun/taskrun_test.go]: nice interface change :D

File: pkg/termination/parse.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+package termination
+
+import (
+	"encoding/json"
+
+	v1alpha1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+)
+
+// ParseMessage parses a termination message as results.
+func ParseMessage(msg string) ([]v1alpha1.PipelineResourceResult, error) {
```
Comment (bobcatfish) [inline pkg/termination/parse.go]: _yaaaasssss_

File: pkg/termination/parse.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	if msg == "" {
+		return nil, nil
+	}
+	var r []v1alpha1.PipelineResourceResult
+	if err := json.Unmarshal([]byte(msg), &r); err != nil {
+		return nil, err
```
Comment (bobcatfish) [inline pkg/termination/parse.go]: another case where i feel like some more error context could help maybe?

File: pkg/termination/parse_test.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+			got, err := ParseMessage(c.msg)
+			if err != nil {
+				t.Fatalf("ParseMessage: %v", err)
+			}
+			if d := cmp.Diff(c.want, got); d != "" {
+				t.Fatalf("ParseMessage(-want,+got): %s", d)
+			}
+		})
+	}
+}
```
Comment (bobcatfish) [inline pkg/termination/parse_test.go]: is there an error case we'd want to test? (e.g. invalid)

Comment (ImJasonH) [reply pkg/termination/parse_test.go]: Sure why not

File: pkg/reconciler/taskrun/taskrun.go (hunk 3 above)
Comment (ImJasonH) [reply pkg/reconciler/taskrun/taskrun.go]: Done.

File: pkg/reconciler/taskrun/taskrun.go (hunk 2 above)
Comment (ImJasonH) [reply pkg/reconciler/taskrun/taskrun.go]: I'll just remove this for now. We're going to need to do more work to generalize resolved resource outputs, and this is one of those places.

File: go.mod (hunk 1 above)
Comment (bobcatfish) [reply go.mod]: this makes me feel a bit :( but we don't have to tackle it here XD

the joys of dependency tools XD

File: pkg/termination/parse_test.go (hunk 7 above)
Comment (bobcatfish) [reply pkg/termination/parse_test.go]: u da best

**Pull request merged**